plugins/inputs/powerdns/powerdns.go

```diff
@@ -49,59 +49,105 @@ func (p *Powerdns) Gather(acc telegraf.Accumulator) error {
 }
 
 func (p *Powerdns) gatherServer(address string, acc telegraf.Accumulator) error {
-	conn, err := net.DialTimeout("unix", address, defaultTimeout)
+	stats_cmd := "show * \n"
+	// fetch general stats
+	stats, err := p.executeCommand(stats_cmd, address)
 	if err != nil {
 		return err
 	}
+	// Process general stats data
+	fields := parseStatsResponse(stats)
+	// Add server socket as a tag
+	tags := map[string]string{"server": address}
+	acc.AddFields("powerdns", fields, tags)
 
-	defer conn.Close()
+	// fetch respsizes stats for calculating histogram metrics
+	respsizes_cmd := "respsizes \n"
+	respsizes, err := p.executeCommand(respsizes_cmd, address)
+	if err != nil {
+		return err
+	}
+	// Process latency stats data
+	latencyBucket := parseRespsizesResponse(respsizes)
+	for latency, count := range latencyBucket {
+		tags["latency"] = strconv.Itoa(int(latency))
+		fields := map[string]interface{}{
+			"resp_size": count,
+		}
+		acc.AddCounter("powerdns", fields, tags)
+	}
+	return nil
+}
 
+func (p *Powerdns) executeCommand(command, address string) (string, error) {
+	conn, err := net.DialTimeout("unix", address, defaultTimeout)
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
 	conn.SetDeadline(time.Now().Add(defaultTimeout))
-
 	// Read and write buffer
 	rw := bufio.NewReadWriter(bufio.NewReader(conn), bufio.NewWriter(conn))
-
 	// Send command
-	if _, err := fmt.Fprint(conn, "show * \n"); err != nil {
-		return nil
+	if _, err := fmt.Fprint(conn, command); err != nil {
+		return "", nil
 	}
 	if err := rw.Flush(); err != nil {
-		return err
+		return "", err
 	}
-
 	// Read data
 	buf := make([]byte, 0, 4096)
 	tmp := make([]byte, 1024)
 	for {
 		n, err := rw.Read(tmp)
 		if err != nil {
 			if err != io.EOF {
-				return err
+				return "", err
 			}
 
 			break
 		}
 		buf = append(buf, tmp[:n]...)
 	}
+	return string(buf), nil
+}
 
-	metrics := string(buf)
-
-	// Process data
-	fields := parseResponse(metrics)
-
-	// Add server socket as a tag
-	tags := map[string]string{"server": address}
-
-	acc.AddFields("powerdns", fields, tags)
-
-	return nil
+func parseRespsizesResponse(rawStats string) map[int64]int64 {
+	// Use 2000 ms as the maximum bucket size
+	maxLatency := int64(2000)
+	latencyBucket := make(map[int64]int64)
+	for _, line := range strings.Split(rawStats, "\n") {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+		parts := strings.Split(line, "\t")
+		if len(parts) < 2 {
+			continue
+		}
+		latency, err := strconv.ParseInt(parts[0], 10, 64)
+		if err != nil {
+			log.Printf("E! powerdns: Error parsing integer for respsizes letency metric [%s]: %s",
```

Comment:
Typo, but I think we should return an error here which would make this stop the gather.